repository-azure/src/main/java/io/aiven/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -31,56 +36,53 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-class AzureClient implements AutoCloseable {
 
-    final Logger logger = LoggerFactory.getLogger(AzureClient.class);
+import static io.aiven.elasticsearch.repositories.CommonSettings.RepositorySettings.MAX_RETRIES;
+
+class AzureClientProvider extends ClientProvider<BlobServiceClient, AzureClientSettings> {
+
+    final Logger logger = LoggerFactory.getLogger(AzureClientProvider.class);
 
     static final String HTTP_USER_AGENT = "aiven-azure-repository";
 
     static final long HTTP_POOL_AWAIT_TERMINATION = 800L;
 
     static final int MAX_HTTP_POOL_SHUTDOWN_ATTEMPTS = 3;
 
-    private final BlobServiceClient blobServiceClient;
-
     /**
      * Thread pool for Azure SDK HTTP client, must be closed explicitly
      */
-    private final ExecutorService httpPoolExecutorService;
-
-    protected AzureClient(final BlobServiceClient blobServiceClient, final ExecutorService httpPoolExecutorService) {
-        this.blobServiceClient = blobServiceClient;
-        this.httpPoolExecutorService = httpPoolExecutorService;
-    }
+    protected volatile ExecutorService httpPoolExecutorService;
 
-    public BlobServiceClient blobServiceClient() {
-        return blobServiceClient;
-    }
-
-    public static AzureClient create(final AzureClientSettings azureClientSettings) {
-        final var httpThreadPoolSettings = azureClientSettings.httpThreadPoolSettings();
-        final var httpClientExecutor = new ThreadPoolExecutor(
+    @Override
+    protected BlobServiceClient buildClient(final AzureClientSettings clientSettings,
+                                            final Settings repositorySettings) {
+        final var httpThreadPoolSettings = clientSettings.httpThreadPoolSettings();
+        httpPoolExecutorService = new ThreadPoolExecutor(
```

Comment:
Interesting, I think the `ThreadPoolExecutor` usage was copied from https://github.com/1184893257/okhttp/blob/master/okhttp/src/main/java/com/squareup/okhttp/Dispatcher.java#L61, where `SynchronousQueue` is being used. The very important difference is that minimum number of threads is set to `0`, which essentially means it is 100% equivalent to `Executors::newCachedThreadPool() ` :-)

Usage of the `LinkedBlockingQueue` is legitimate but I think we should have it bounded to a reasonable value (may be configurable as well as part of `httpThreadPoolSettings`.


```
executorService = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60, TimeUnit.SECONDS,
          new SynchronousQueue<Runnable>(), Util.threadFactory("OkHttp Dispatcher", false));
```

Reply:
True the default value should be `1`. Regarding queue size, I think default value `1000` and min value `10` should be enough. What do you think?

Reply:
> > True the default value should be 1
> 
> The behavior of the ThreadPoolExecutor is confusing but here is the brief summary: the pool starts with `min` threads, **fills up the queue** and only than scales to `max` threads. So setting it to 1 is not what you want, it is better to have `min` == `max`, as we have right now (https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html for the reference).

No we set it to 1 by default. Changed to min == max